### ISSUE TITLE
Error on initializing context, only on OpenCL platform

### ISSUE THREAD

**Issue by orioncohen** [openmm/openmm#3380]
Hello, 

I am encountering a very confusing error when trying to run OpenMM with OpenCL, some troubleshooting pointers would be incredibly helpful.

## The Setup

I am debugging some OpenMM code on one node of the NERSC Perlmutter supercomputer, which has 4 A100 Ampere GPU's. 

I am running two scripts with different setup procedures. Both scripts work locally on my laptop and work remotely with CPUs, but one of them fails when run remotely with OpenCL. It returns a `openmm.OpenMMException: Error initializing context: clCreateContext (-33)`. The full traceback is below.

```
Traceback (most recent call last):
  File "/global/u1/o/oac/openmm/openmm-testing/pymatgen_sims/pymatgen_sim_EA.py", line 80, in <module>
    sim = openmm.app.Simulation(
  File "/global/homes/o/oac/miniconda3/envs/openmm/lib/python3.9/site-packages/openmm/app/simulation.py", line 103, in __init__
    self.context = mm.Context(self.system, self.integrator, platform)
  File "/global/homes/o/oac/miniconda3/envs/openmm/lib/python3.9/site-packages/openmm/openmm.py", line 3951, in __init__
    _openmm.Context_swiginit(self, _openmm.new_Context(*args))
openmm.OpenMMException: Error initializing context: clCreateContext (-33)
```

This error happens when I try to instantiate a simulation from a pre-made `system`, `topology`, and `integrator`. I am also specifying a `platform` and `platformProperties` . My setup looks something like this:

```python
properties = {"DeviceIndex": f"{0}"} 
platform = Platform.getPlatformByName("OpenCL")

# this is bespoke code for setting up the system, it is quite long, and I suspect that my error is somewhere in here
system = input_set.inputs['system.xml'].get_system()
topology = input_set.inputs['topology.pdb'].get_topology()
integrator = LangevinMiddleIntegrator(
    300 * kelvin, 1 / picosecond, 0.001 * picoseconds
)
state = input_set.inputs['state.xml'].get_state()

# this call is throwing an error.
sim = openmm.app.Simulation(
    topology,
    system,
    integrator,
    platform=platform,
    platformProperties=properties
)
```

Any debugging advice would be appreciated. I am not looking for a full a solution. Just some advice on what to try.

## on another note:

My benchmarks have found that for a system of ~6000 atoms, I get a throughput of ~300 ns/day on a single GPU, which translates to 1200 ns/day of total throughput per node. Pretty dang fast! I'd be happy to run some more formal benchmarks if the OpenMM devs are interested in the data.

**Comment by peastman** [comment]
Error -33 is `CL_INVALID_DEVICE`.  What happens if you leave out the `DeviceIndex` property and let it pick a device itself?

**Comment by orioncohen** [comment]
Thanks for the response, knowing the error type is helpful, but sadly that fix doesn't work. 

I have managed to narrow down the error a bit though. It seems to come from duplicate calls to create a simulation. I have written a function, `_smiles_to_simulation` that constructs a system, topology, and integrator and then creates a simulation, something like this:

```python
    platform = Platform.getPlatformByName('OpenCL')
    simulation = Simulation(topology, system, integrator, platform=platform)
    simulation.context.setPositions(coordinates)
    return simulation
```
This function works time the first time I call it, but a second call will fail.

```python
# succeeds
simulation = _smiles_to_simulation(
    [TFEA, FEC, Li, PF6],   # these are smiles
    [512, 88, 62, 62],        # these are amounts
)

# fails
simulation = _smiles_to_simulation(
    [TFEA, FEC, Li, PF6],   # these are smiles
    [512, 88, 62, 62],        # these are amounts
)
```

Do different instances of `Simulation` compete for the same platform? That seems to be whats happening.

Unfortunately this doesn't completely solve my problem, because I have another function like `_smiles_to_simulation` that fails even if I only call it once, but this is some insight.

**Comment by orioncohen** [comment]
Ok, I can't say exactly what the bug is but I seem to have maneuvered around it.

My previous workflow was instantiating a `Context` before instantiating the `Simulation`. This seems to have somehow caused `Simulation` to run into an error when internally creating a `Context`. I am not sure why this was only an issue on `OpenCL` and not `CPU`, but I am glad it is resolved. I am closing this issue now, hopefully it will help someone in the future!

tl;dr if you see this error, make sure you are only ever instantiating one context!

**Comment by peastman** [comment]
This probably means the GPU is set to exclusive mode.  In that mode, it only allows one context to exist on it at a time.  If you try to create a second one, you get an error.  You can check what mode it's set in with `nvidia-smi`.  You can also use it to change the mode, but since you're on a cluster you don't administer you probably don't have permission to.

Before creating a new Simulation, first delete any existing Context or Simulation (which internally creates a Context):

```python
delete context
```

Does that work?

**Comment by orioncohen** [comment]
Ahh, that's great to know. You are right, they are all exclusive. 

I tried calling `del context` and explicitly calling `gc.collect()` afterwards (as you suggested in issue #1501), but that did not work. I am only instantiating one `Context` and it is out of scope by the time I generate a `Simulation`, so I am not sure why the GPU reservation is so persistent.

**Comment by jchodera** [comment]
Can you share your complete code snippet here? The garbage collection with GPIU resource management is actually kind of tricky.

**Comment by orioncohen** [comment]
Here is what I did:

```python
context = Context(system, integrator)
context.setPositions(coordinates)
state = context.getState(getPositions=True)
del context
from gc import collect
collect()
```

After this, context passes out of scope before a simulation is instantiated. At this point in the code, I am not providing any platform to Context.

**Comment by jchodera** [comment]
@orioncohen : Apologies for the delay. I was hoping you could share a more complete code snippet that includes the entire context creation, simulation, and cleanup?.

For NVIDIA GPUs, if you're not using the "default" (shared) mode, you can only have one `Context` open at a time. The `Simulation` object has a `Context` internally, and I'm not certain whether `del simulation` is sufficient to also delete the `Context` it contains, so you may have to `del simulation.context` directly before creating the next `Simulation` object. But the cycle should look like
* Create a `Context` (either directly or via a `Simulation` object)
* Set box vectors and positions 
* Run some dynamics
* Get `State` for properties of interest
* `del context` to clean up the context and safely be able to create a new simulation

It will be easier to point out exactly where things are happening out of order if you can share a bit more code for the entire routine that handles this cycle.
